File: src/main/java/rak/planetGenerator/factories/GeneratorFactory.java
package rak.planetGenerator.factories;

import rak.planetGenerator.generators.BiomeGenerator;
import rak.planetGenerator.generators.HeightMapGenerator;
import rak.planetGenerator.generators.PercipitationMapGenerator;
import rak.planetGenerator.generators.PlanetGenerator;
import rak.planetGenerator.generators.TemperatureMapGenerator;
import rak.planetGenerator.generators.arguments.GenerationArguments;
import rak.planetGenerator.model.PlanetType;

public class GeneratorFactory {
	
	public PlanetGenerator createGenerator(GenerationArguments defaultArguments, GenerationArguments overrideArguments){
		HeightMapGenerator heightGenerator = new HeightMapGenerator();
		defaultArguments.apply(heightGenerator);
		overrideArguments.apply(heightGenerator);
		
		TemperatureMapGenerator temperatureGenerator = new TemperatureMapGenerator();
		defaultArguments.apply(temperatureGenerator);
		overrideArguments.apply(temperatureGenerator);
		
		PercipitationMapGenerator percipitationMapGenerator = new PercipitationMapGenerator();
		defaultArguments.apply(percipitationMapGenerator);
		overrideArguments.apply(percipitationMapGenerator);
		
		BiomeGenerator biomeGenerator = new BiomeGenerator();
		defaultArguments.apply(biomeGenerator);
		overrideArguments.apply(biomeGenerator);
		
		return new PlanetGenerator(heightGenerator, temperatureGenerator, percipitationMapGenerator, biomeGenerator);
	}
	
	public PlanetGenerator createGeneratorByType(PlanetType type, GenerationArguments overrideArguments){
		GenerationArguments defaultArguments = type.getDefaultArguments();
		return createGenerator(defaultArguments, overrideArguments);
	}
}

File: src/main/java/rak/planetGenerator/factories/GeneratorFactory.kt
package rak.planetGenerator.factories

import rak.planetGenerator.generators.*
import rak.planetGenerator.generators.arguments.GenerationArguments
import rak.planetGenerator.model.PlanetType

class GeneratorFactory {

    fun createGenerator(defaultArguments: GenerationArguments, overrideArguments: GenerationArguments): PlanetGenerator {
        val heightGenerator = HeightMapGenerator()
        defaultArguments.apply(heightGenerator)
        overrideArguments.apply(heightGenerator)

        val temperatureGenerator = TemperatureMapGenerator()
        defaultArguments.apply(temperatureGenerator)
        overrideArguments.apply(temperatureGenerator)

        val precipitationMapGenerator = PrecipitationMapGenerator()
        defaultArguments.apply(precipitationMapGenerator)
        overrideArguments.apply(precipitationMapGenerator)

        val biomeGenerator = BiomeGenerator()
        defaultArguments.apply(biomeGenerator)
        overrideArguments.apply(biomeGenerator)

        return PlanetGenerator(heightGenerator, temperatureGenerator, precipitationMapGenerator, biomeGenerator)
    }

    fun createGeneratorByType(type: PlanetType, overrideArguments: GenerationArguments): PlanetGenerator {
        val defaultArguments = type.defaultArguments
        return createGenerator(defaultArguments, overrideArguments)
    }
}
